Feature/native security policygen regen

mvo5 opened the pull request:
[This branch is based on feature/native-security-policygen-compare, compare e.g. with https://github.com/mvo5/snappy/compare/feature/native-security-policygen-compare...mvo5:feature/native-security-policygen-regen]

This branch adds "sudo snappy policygen --regenerate-all" that will check for any outdated policy and re-generates them. The idea is to use it in a systemd unit at bootup.

mvo5 commented:
Thanks for your earlier irc  irc feedback! I added "c8fc24d" which will only re-generate the policies if any version of apparmor or ubuntu-core-security-* changes. As of livecd-rootfs 2.351 this file is part of the base system now.

jdstrand commented:
I'm going to pull this in as is since in the interest of easing development on the branch, however, there are a few problems:
* sd.mergeAppArmorSecurityOverrides(&hwaccessOverrides) results in a nil pointer dereference with a test snap I have that uses:
 - name: bin/testme-override
   description: "testme-override - binary"
   security-override:
     seccomp:
       syscalls: ["foo", "bar"]
* snappy remove doesn't seem to remove all of the old profiles, and purge doesn't either
* snappy hw-assign doesn't work right with security-policy

On the plus side, this branch seems to resolve the issue I had before with --compare.

mvo5 commented:
On Mon, Nov 09, 2015 at 03:34:34PM -0800, jdstrand wrote:
> I'm going to pull this in as is since in the interest of easing development on the branch, however, there are a few problems:
> * sd.mergeAppArmorSecurityOverrides(&hwaccessOverrides) results in a nil pointer dereference with a test snap I have that uses:
>  - name: bin/testme-override
>    description: "testme-override - binary"
>    security-override:
>      seccomp:
>        syscalls: ["foo", "bar"]

Thanks! This is fixed in git and I also added more testcases to ensure
this won't break again.

> * snappy remove doesn't seem to remove all of the old profiles, and
>   purge doesn't either

How can I reproduce this? I did:
"""
(amd64)ubuntu@localhost:~$ sudo ./snappy install hello-world
Installing hello-world
Starting download of hello-world
21.60 KB / 21.60 KB [===================================] 100.00 %
549.15 KB/s 0
Done
Name          Date       Version Developer
ubuntu-core   2015-11-04 227     ubuntu
hello-world   2015-11-10 1.0.18  canonical
generic-amd64 2015-10-21 1.4     canonical
Reboot to use the new ubuntu-core.
(amd64)ubuntu@localhost:~$ ls -l /var/lib/snappy/apparmor/profiles/
total 48
-rw-r--r-- 1 root root 7975 Nov 10 10:21
hello-world.canonical_echo_1.0.18
-rw-r--r-- 1 root root 7972 Nov 10 10:21
hello-world.canonical_env_1.0.18
-rw-r--r-- 1 root root 7975 Nov 10 10:21
hello-world.canonical_evil_1.0.18
-rw-r--r-- 1 root root 7969 Nov 10 10:21
hello-world.canonical_sh_1.0.18
-rw-r--r-- 1 root root 7984 Nov 10 10:21
hello-world.canonical_showdev_1.0.18
-rw-r--r-- 1 root root 7978 Nov 10 10:21
hello-world.canonical_usehw_1.0.18
(amd64)ubuntu@localhost:~$ sudo ./snappy remove hello-world
Removing hello-world
(amd64)ubuntu@localhost:~$ ls -l /var/lib/snappy/apparmor/profiles/
total 0
"""
(same in seccomp). So it seems to work (unless I miss something).

Did you maybe not look into the new "/var/lib/snappy" prefix dir?

> * snappy hw-assign doesn't work right with security-policy

Indeed, I added this now.

> On the plus side, this branch seems to resolve the issue I had before with --compare.

Yay!
